tools/sst_rewrite.cc

```diff
@@ -0,0 +1,353 @@
+
+    return;
+
+}   // LDbTable::LDbTable
+
+
+LDbTable::~LDbTable()
+{
+    m_IsOpen=false;
+    delete m_TablePtr;
+    m_TablePtr=NULL;
+    delete m_FilePtr;
+    m_FilePtr=NULL;
+    m_FileSize=0;
+}   // LDbTable::~LDbTable
+
+
+leveldb::Iterator *
+LDbTable::NewIterator()
+{
+    leveldb::Iterator * ret_ptr(NULL);
+
+    if (m_IsOpen)
+    {
+        leveldb::ReadOptions read_options;
+
+        read_options.fill_cache=false;
+        ret_ptr=m_TablePtr->NewIterator(read_options);
+    }   // if
+
+    return(ret_ptr);
+
+}   // LDbTable::NewIterator
+
+
+int
+main(
+    int argc,
+    char ** argv)
+{
+    bool error_seen, running, compare_files;
+    char ** cursor;
+
+    compare_files=false;
+    error_seen=false;
+    running=true;
+
+    // Options: needs filter & total_leveldb_mem initialized
+    leveldb::Options options;
+    options.filter_policy=leveldb::NewBloomFilterPolicy2(16);
+    options.total_leveldb_mem=(512 << 20);
+
+    // testing
+//    options.block_size=16 << 10;
+
+    for (cursor=argv+1; NULL!=*cursor && running; ++cursor)
+    {
+        // option flag?
+        if ('-'==**cursor)
+        {
+            char flag;
+
+            flag=*((*cursor)+1);
+            switch(flag)
+            {
+                //case 'b':  options.block_size=4096; break;
+                case 'c':  compare_files=true; break;
+                case 'w':  compare_files=false; break;
+                default:
+                    fprintf(stderr, " option \'%c\' is not valid\n", flag);
+                    command_help();
+                    running=false;
+                    error_seen=true;
+                    break;
+            }   // switch
+        }   // if
+
+        // sst file
+        else
+        {
+            std::string fname;
+            fname=*cursor;
+
+            // do a rewrite
+            if (!compare_files)
+            {
+                leveldb::WritableFile *outfile = NULL;
+                leveldb::Status s;
+                leveldb::Iterator *it;
+                leveldb::TableBuilder * builder = NULL;
+
+                LDbTable in_file(options, fname);
+
+                if (in_file.GetStatus().ok())
+                {
+                    it=in_file.NewIterator();
+
+                    fname.append(".new");
+                    s = options.env->NewWritableFile(fname, &outfile,
+                                                     options.env->RecoveryMmapSize(&options));
+                    if (s.ok())
+                        builder = new leveldb::TableBuilder(options, outfile);
+                    else
+                    {
+                        // Table::Open failed on file "fname"
+                        fprintf(stderr, "%s: NewWritableFile failed (%s)\n",
+                                fname.c_str(), s.ToString().c_str());
+                        error_seen=true;
+                    }   // else
+
+                    for (it->SeekToFirst(); it->Valid() && s.ok(); it->Next())
```

Comment:
I recommend adding a call to builder->status.ok() in the conditional section of the for loop. Similarly, I suggest adding the same check on line 216.